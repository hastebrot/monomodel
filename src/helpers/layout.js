import React, { Fragment, useEffect, useState } from "react"
import { Box, ThemeProvider } from "fannypack"
import { produce } from "immer"

export const UseGridPositionsDemo = () => {
  const { items, setItems, toItemProps, containerProps } = useGridPositions({
    containerSize: { w: 2, h: 4 },
    initialItems: [
      { id: 0, text: "0", bounds: { y: 0, x: 0, h: 1, w: 2 } },
      { id: 1, text: "1", bounds: { y: 1, x: 0, h: 1, w: 1 } },
      { id: 2, text: "2", bounds: { y: 1, x: 1, h: 1, w: 1 } },
      { id: 3, text: "3", bounds: { y: 2, x: 1, h: 1, w: 1 } },
      { id: 4, text: "4", bounds: { y: 3, x: 0, h: 1, w: 1 } },
    ],
  })

  useEffect(() => {
    setItems(
      produce(items => {
        moveArrayItemInplace(items, 0, 2)
      })
    )
  }, [])

  return (
    <ThemeProvider>
      <Box {...containerProps}>
        {items.map(item => (
          <Fragment key={item.id}>
            <Box
              padding="10px"
              border="1px solid #ccc"
              backgroundColor="#fff"
              {...toItemProps(item.id)}
            >
              {item.text}
            </Box>
          </Fragment>
        ))}
      </Box>
    </ThemeProvider>
  )
}

export const useGridPositions = ({ initialItems, containerSize }) => {
  const [items, setItems] = useState(initialItems)
  const verticalScaleFactor = 50
  return {
    containerProps: {
      position: "relative",
      width: "100%",
      height: `${containerSize.h * verticalScaleFactor}px`,
    },
    items,
    setItems,
    toItemProps: id => {
      const item = items.find(item => item.id === id)
      const itemBounds = {
        x: (item.bounds.x / containerSize.w) / (item.bounds.w / containerSize.w),
        y: (item.bounds.y / containerSize.h) * (containerSize.h * verticalScaleFactor),
        w: item.bounds.w / containerSize.w,
        h: item.bounds.h * verticalScaleFactor,
      }
      return {
        position: "absolute",
        width: `${itemBounds.w * 100}%`,
        height: `${itemBounds.h}px`,
        transform: `translate(${itemBounds.x * 100}%, ${itemBounds.y}px)`,
      }
    },
  }
}

export const UseGridColumnsDemo = () => {
  const CONTAINER_COLUMNS = "repeat(2, 1fr)"
  const FULL_COLUMN = "1 / span 2"
  const HALF_COLUMN = "span 1"
  const LEFT_COLUMN = "1 / span 1"
  const RIGHT_COLUMN = "2 / span 1"

  const { items, setItems, toItemProps, containerProps } = useGridColumns({
    containerColumns: CONTAINER_COLUMNS,
    initialItems: [
      { id: 0, text: "0", columns: FULL_COLUMN },
      { id: 1, text: "1", columns: HALF_COLUMN },
      { id: 2, text: "2", columns: HALF_COLUMN },
      { id: 3, text: "3", columns: RIGHT_COLUMN },
      { id: 4, text: "4", columns: LEFT_COLUMN },
    ],
  })

  useEffect(() => {
    setItems(
      produce(items => {
        moveArrayItemInplace(items, 0, 2)
      })
    )
  }, [])

  return (
    <ThemeProvider>
      <Box {...containerProps}>
        {items.map(item => (
          <Fragment key={item.id}>
            <Box
              padding="10px"
              border="1px solid #ccc"
              backgroundColor="#fff"
              {...toItemProps(item.id)}
            >
              {item.text}
            </Box>
          </Fragment>
        ))}
      </Box>
    </ThemeProvider>
  )
}

export const useGridColumns = ({ initialItems, containerColumns }) => {
  const [items, setItems] = useState(initialItems)
  return {
    containerProps: {
      display: "grid",
      gridTemplateColumns: containerColumns,
      gridColumnGap: "16px",
      gridRowGap: "16px",
    },
    items,
    setItems,
    toItemProps: id => {
      const item = items.find(item => item.id === id)
      return {
        gridColumn: item.columns,
      }
    },
  }
}

export const moveArrayItemInplace = (array, indexFrom, indexTo) => {
  array.splice(
    indexTo < 0 ? array.length + indexTo : indexTo,
    0,
    array.splice(indexFrom, 1)[0]
  )
}
