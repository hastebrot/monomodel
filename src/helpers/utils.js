import React from "react"

export const check = (condition, message = () => "condition is false") => {
  if (!condition) {
    throw new Error(message())
  }
}

export const checkNotNil = (value, message = () => "value is nil") => {
  if (isNil(value)) {
    throw new Error(message())
  }
}

export const isNil = value => {
  return value === null || value === undefined
}

export const PrettyCode = ({ value, indent = 2, ...otherProps }) => {
  return (
    <pre
      style={{
        margin: "0",
        padding: "0",
        color: "#000",
        fontSize: "13px",
        whiteSpace: "pre-wrap",
        wordWrap: "break-word",
      }}
      {...otherProps}
    >
      <code>{pretty(value, indent)}</code>
    </pre>
  )
}

export const pretty = (value, indent = 2) =>
  JSON.stringify(value, undefined, indent)

export const pretty1 = value =>
  pretty(value, 1)
    .split("\n")
    .map(it => it.trim())
    .join(" ")

export const pretty0 = value => pretty(value, 0)
