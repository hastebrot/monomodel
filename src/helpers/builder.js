import ptr from "json-ptr"
import { schemaWalk } from "@cloudflare/json-schema-walker"
import graphlib, { Graph } from "graphlib"
import { get, set, pick, concat, cloneDeep } from "lodash"
import { fieldset, field } from "./model"
import { pretty1 } from "./utils"

// TODO(hastebrot): buildModel() + guestPointers

export const buildModel = (schema, metaModel, document) => {
  // console.log(ptr.flatten(schema))

  const listPointerSegments = (pointer, graph) => {
    let currentPointer = pointer
    const pointerSegments = []
    while (currentPointer) {
      const node = graph.node(currentPointer)
      const parentNode = graph.node(graph.parent(currentPointer))
      pointerSegments.unshift({
        segment: node.pointerSegment,
        parentType: parentNode ? parentNode.schema.type : null,
      })
      currentPointer = graph.parent(currentPointer)
    }
    return pointerSegments
  }

  const buildRegistry = graph => {
    const registry = {}
    for (const selfPointer of traverseSchemaGraph(graph)) {
      const parentPointer = graph.parent(selfPointer)
      const selfNode = graph.node(selfPointer)
      const parentNode = graph.node(parentPointer)
      registry[selfPointer] = {
        segment: selfNode.pointerSegment,
        type: parentNode ? parentNode.schema.type : null,
        parentPointer,
      }
    }
    return registry
  }

  const listParents = (pointer, registry) => {
    let currentPointer = pointer
    const parents = []
    while (currentPointer) {
      parents.unshift(currentPointer)
      currentPointer = registry[currentPointer]
        ? registry[currentPointer].parentPointer
        : null
    }
    return parents.map(pointer =>
      registry[pointer] ? registry[pointer].segment : null
    )
  }

  const pointer = "#/properties/orderItems/items/properties/productNumber"
  const graph = buildSchemaGraph(schema)
  console.log(writeSchemaGraph(graph))
  console.log(listPointerSegments(pointer, graph))

  const registry = buildRegistry(graph)
  console.log(listParents(pointer, registry))

  const rewriteSegments = (pointer, graph) => {
    const rewriteObjectSegment = pointerSegment => {
      return ptr.decodePointer(pointerSegment).slice(-1)[0]
    }
    const rewriteArraySegment = pointerSegment => {
      return "0"
    }

    const s = listPointerSegments(pointer, graph)
    const t = s.map(seg => {
      if (seg.parentType === "object") {
        return rewriteObjectSegment(seg.segment)
      }
      if (seg.parentType === "array") {
        return rewriteArraySegment(seg.segment)
      }
      return null
    })
    const path = ptr.encodePointer(t.slice(1))
    return path ? "#" + path : "#/"
  }

  const model = cloneDeep(metaModel)

  for (const fieldset of model.children) {
    fieldset.docPointer = rewriteSegments(fieldset.pointer, graph)
    if (fieldset.children) {
      for (const field of fieldset.children) {
        field.docPointer = rewriteSegments(field.pointer, graph)
      }
    }
  }

  model.children.splice(3, 0, {
    type: "context",
    title: "context start: repeatable fieldsets",
  })
  model.children.splice(5, 0, {
    type: "context",
    title: "context end: repeatable fieldsets",
  })

  return model

  // const model = []
  // const context = {}
  // const lastIndex = metaModel.length - 1
  // let index = 0
  // while (index <= lastIndex) {
  //   const e = metaModel[index]
  //   const p = e.self.p
  //   const t = e.self.t
  //   if (e.parent && e.parent.p in context) {
  //     const lastSubIndex = context[e.parent.p].size - 1
  //     let subIndex = 0
  //     while (subIndex <= lastSubIndex) {
  //       context.indices = [subIndex]
  //       const q = resolutions[p](context)
  //       const v = ptr.get(document, q)
  //       model.push({ p, q, t })
  //       subIndex += 1
  //     }
  //   } else {
  //     const q = resolutions[p](context)
  //     const v = ptr.get(document, q)
  //     model.push({ p, q, t })
  //     if (t === "array") {
  //       context[p] = { size: v.length }
  //     }
  //   }
  //   index += 1
  // }
}

/**
 * Builds a flat meta model using a json schema.
 */
export const buildMetaModelFlat = schema => {
  const registry = {
    fieldsetIndex: -1,
    fieldIndex: -1,
    currentParentPointer: null,
  }
  const model = fieldset("array")
  traverseSchema(schema, (object, parentObject) => {
    const pointer = toPointer(concat(parentObject.path, object.path))
    const parentPointer = toPointer(parentObject.path)
    const type = object.schema.type

    if (
      !isFieldsetType(type) &&
      parentPointer !== registry.currentParentPointer
    ) {
      registry.fieldsetIndex += 1
      registry.fieldIndex = -1
      registry.currentParentPointer = parentPointer

      const objectPathSegments = []
      objectPathSegments.push(`children[${registry.fieldsetIndex}]`)
      const objectPath = toObjectPath(concat(...objectPathSegments))
      // console.log(pretty1([objectPath, parentPointer]))

      const payload = pick(parentObject.schema, ["title", "description"])
      const node = fieldset(parentObject.schema.type, {
        pointer: parentPointer,
        ...payload,
      })
      set(model, objectPath, node)
    }

    if (isFieldsetType(type)) {
      registry.fieldsetIndex += 1
      registry.fieldIndex = -1
      registry.currentParentPointer = pointer
    } else {
      registry.fieldIndex += 1
    }

    const objectPathSegments = []
    objectPathSegments.push(`children[${registry.fieldsetIndex}]`)
    if (!isFieldsetType(type)) {
      objectPathSegments.push(`children[${registry.fieldIndex}]`)
    }
    const objectPath = toObjectPath(concat(...objectPathSegments))
    // console.log(pretty1([objectPath, pointer]))

    if (isFieldsetType(type)) {
      const payload = pick(object.schema, ["title", "description"])
      const node = fieldset(type, { pointer, ...payload })
      set(model, objectPath, node)
    } else {
      const payload = pick(object.schema, ["title", "description"])
      const node = field(type, { pointer, ...payload })
      set(model, objectPath, node)
    }
  })
  return model
}

/**
 * Builds a nested meta model using a json schema.
 */
export const buildMetaModelNested = schema => {
  const registry = {}
  const model = {}
  traverseSchema(schema, (object, parentObject) => {
    const pointer = toPointer(concat(parentObject.path, object.path))
    const parentPointer = toPointer(parentObject.path)
    const type = object.schema.type

    if (isRootPointer(pointer, parentPointer)) {
      registry[pointer] = {
        parentPointer: null,
        path: null,
        index: 0,
        nextChildIndex: 0,
      }
    } else {
      registry[pointer] = {
        parentPointer,
        path: object.path,
        index: registry[parentPointer].nextChildIndex,
        nextChildIndex: 0,
      }
      registry[parentPointer].nextChildIndex += 1
    }

    const rewritePathSegment = ({ path, index }) => {
      // return path || []
      return path ? ["children", index] : []
    }

    const objectPathSegments = []
    let currentEntry = registry[pointer]
    while (currentEntry) {
      objectPathSegments.unshift(rewritePathSegment(currentEntry))
      currentEntry = registry[currentEntry.parentPointer]
    }
    const objectPath = toObjectPath(concat(...objectPathSegments))
    const rootObjectPath = objectPath ? "root." + objectPath : "root"
    // console.log(pretty1([objectPath, pointer]))

    if (isFieldsetType(type)) {
      const payload = pick(object.schema, ["title", "description"])
      const node = fieldset(type, { pointer, ...payload })
      set(model, rootObjectPath, node)
    } else {
      const payload = pick(object.schema, ["title", "description"])
      const node = field(type, { pointer, ...payload })
      set(model, rootObjectPath, node)
    }
  })
  return model.root
}

export const isFieldsetType = type => {
  return type === "object" || type === "array"
}

export const isRootPointer = (pointer, parentPointer) => {
  return pointer === parentPointer
}

export const toPointer = pathSegments => {
  return `#/${pathSegments.join("/")}`
}

export const toObjectPath = pathSegments => {
  return pathSegments.join(".")
}

export const traverseSchema = (schema, visitFn) => {
  schemaWalk(schema, (schemaObject, path, parentSchemaObject, parentPath) => {
    const object = { schema: schemaObject, path: path }
    const parentObject = { schema: parentSchemaObject, path: parentPath }
    visitFn(object, parentObject)
  })
}

const ROOT_POINTER = "#/"

export const buildSchemaGraph = schema => {
  const isRootPointer = pointer => {
    return pointer === ROOT_POINTER
  }
  const withRoot = pointer => (pointer ? "#" + pointer : ROOT_POINTER)
  const graph = new Graph({ directed: true, compound: true })
  schemaWalk(schema, (selfSchema, selfPath, parentSchema, parentPath) => {
    const schema = selfSchema
    const pointer = withRoot(ptr.encodePointer(concat(parentPath, selfPath)))
    const parentPointer = withRoot(ptr.encodePointer(concat(parentPath)))
    const pointerSegment = ptr.encodePointer(selfPath)
    graph.setNode(pointer, {
      schema,
      pointer,
      pointerSegment,
    })
    if (!(isRootPointer(parentPointer) && isRootPointer(pointer))) {
      graph.setEdge(parentPointer, pointer)
    }
    if (!isRootPointer(pointer)) {
      graph.setParent(pointer, parentPointer)
    }
  })
  return graph
}

export const traverseSchemaGraph = (graph, root = ROOT_POINTER) =>
  graphlib.alg.preorder(graph, root)

export const readSchemaGraph = json => graphlib.json.read(json)
export const writeSchemaGraph = graph => graphlib.json.write(graph)
