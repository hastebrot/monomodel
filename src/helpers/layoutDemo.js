import React, { Fragment, useState, useEffect, useContext } from "react"
import { produce } from "immer"
import { get } from "lodash-es"
import { Box, Flex, Heading, Text, InputField } from "fannypack"
import { List, arrayMove } from "react-movable"
import GridLayout from "react-grid-layout"
import "react-grid-layout/css/styles.css"
import "react-resizable/css/styles.css"
import {
  FormRoot,
  FormContext,
  ContextProvider,
  defaultFormContext,
  defaultLocalTheme,
} from "../components/FormRoot"
import { object, array, string, boolean, number } from "../helpers/model"
import { buildFlatModel } from "../helpers/builder"
import { pretty } from "../helpers/utils"

export const FormDemoOne = () => {
  const formModel = buildFlatModel(orderSchema)
  const formRegistry = {
    FormArray: EditorFormArray,
    FormPart: EditorFormPart,
  }
  return <FormViewer model={formModel} registry={formRegistry} />
}

export const FormDemoTwo = () => {
  const formModel = buildFlatModel(orderSchema)
  const formRegistry = {
    FormPart: EditorFormPart,
  }
  const gridOptions = {
    width: 600,
    cols: 6,
    rowHeight: 60,
    margin: [16, 16],
    containerPadding: [0, 0],
    compactType: "horizontal",
    layout: [
      { i: "a", x: 0, y: 0, w: 1, h: 1, static: true },
      { i: "b", x: 1, y: 0, w: 3, h: 1, minW: 2, maxW: 4 },
      { i: "c", x: 4, y: 0, w: 1, h: 1 },
    ],
  }
  return (
    <GridLayout {...gridOptions}>
      {gridOptions.layout.map(item => (
        <Flex key={item.i}>
          <Box background="#ccc" flex="1" padding="major-1">
            {item.i}
          </Box>
        </Flex>
      ))}
    </GridLayout>
  )
  // return <FormViewer model={formModel} registry={formRegistry} />
}

export const FormDemoThree = () => {
  const formModel = buildFlatModel(orderSchema)
  const formRegistry = {
    FormPart: EditorFormPart,
  }
  return <FormViewer model={formModel} registry={formRegistry} />
}

export const FormViewer = ({
  model = null,
  registry = null,
  ...otherProps
}) => {
  const [formModel, setFormModel] = useState(model)
  const localTheme = defaultLocalTheme()
  const formContext = defaultFormContext(
    produce(context => {
      context.registry = { ...context.registry, ...registry }
      context.formModel = formModel
      context.setFormModel = setFormModel
    })
  )
  return (
    <ContextProvider formContext={formContext} localTheme={localTheme}>
      <Box
        _paddingTop="major-2"
        paddingLeft="major-2"
        paddingRight="major-2"
        color="#121212"
        {...otherProps}
      >
        {formModel && <FormRoot rootModel={formModel} />}
      </Box>
    </ContextProvider>
  )
}

export const EditorFormArray = ({ items, path }) => {
  const { registry, queryComponent, setFormModel } = useContext(FormContext)
  return (
    <List
      lockVertically
      _transitionDuration={0}
      values={items}
      onChange={({ oldIndex, newIndex }) => {
        setFormModel(
          produce(formModel => {
            const array = get({ root: formModel }, path)
            arrayMoveMutate(array, oldIndex, newIndex)
          })
        )
      }}
      renderList={({ children, props }) => {
        return <Box {...props}>{children}</Box>
      }}
      renderItem={({ value, index, props }) => {
        return (
          <Box {...props} key={`${value.pointer}@${index}`}>
            <Box
              use={queryComponent(registry, "FormObject", value)}
              node={value}
              path={`${path}[${index}]`}
            />
          </Box>
        )
      }}
    />
  )
}

export const EditorFormPart = ({ node, path, children }) => {
  const context = useContext(FormContext)

  const header = node.pointer ? (
    <Text>
      {node.title} <Text use="small">{node.pointer}</Text>
    </Text>
  ) : null

  const rowPadding = paddingPx => {
    return {
      paddingLeft: `${paddingPx}`,
      paddingRight: `${paddingPx}`,
      marginLeft: `-${paddingPx}`,
      marginRight: `-${paddingPx}`,
    }
  }

  if (node.type === "fieldset") {
    return (
      <DropBox>
        <Box
          {...rowPadding("16px")}
          border={`1px solid ${node.pointer ? "red" : "red"}`}
          cursor="move"
        >
          {header && (
            <Box {...rowPadding("16px")} paddingBottom="8px">
              <Heading use="h2" paddingTop="8px" paddingBottom="8px">
                <Fragment>{header}</Fragment>
              </Heading>
            </Box>
          )}
          {children && <Box {...rowPadding("16px")}>{children}</Box>}
        </Box>
      </DropBox>
    )
  }
  if (node.type === "field") {
    return (
      <DragBox>
        <Box marginBottom="16px" border="1px solid blue">
          <InputField
            label={node.pointer.split("/").slice(-1)[0]}
            defaultValue={node.pointer}
            readOnly
          />
        </Box>
      </DragBox>
    )
  }
  throw new Error("node.type handler not defined")
}

export const DropBox = ({ children, name, ...otherProps }) => {
  return <Box {...otherProps}>{children}</Box>
}

export const DragBox = ({ children, name, ...otherProps }) => {
  return (
    <Box draggable {...otherProps}>
      {children}
    </Box>
  )
}

const arrayMoveMutate = (array, indexFrom, indexTo) => {
  array.splice(
    indexTo < 0 ? array.length + indexTo : indexTo,
    0,
    array.splice(indexFrom, 1)[0]
  )
}

export const orderSchema = object({
  $schema: "http://json-schema.org/schema#",
  title: "order",
  properties: {
    orderNumber: string(),
    orderDate: string(),
    customer: object({
      title: "customer",
      properties: {
        customerNumber: string(),
        firstName: string(),
        lastName: string(),
      },
    }),
  },
})
