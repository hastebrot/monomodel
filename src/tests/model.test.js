import React from "react"
import renderer from "react-test-renderer"
import { fetchSchemaPaths, orderSchema } from "../helpers/model"

describe("model", () => {
  it("can render react component", () => {
    // when:
    const component = renderer.create(<div>foo</div>)

    // then:
    expect(component.toJSON()).toMatchSnapshot()
  })

  it("can walk json schema", () => {
    // given:
    const schema = orderSchema

    // when:
    const schemaPaths = fetchSchemaPaths(schema)

    // then:
    expect(schemaPaths).toMatchSnapshot()
  })
})
