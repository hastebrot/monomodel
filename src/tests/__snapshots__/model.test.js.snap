// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model can render react component 1`] = `
<div>
  foo
</div>
`;

exports[`model can walk json schema 1`] = `
Array [
  Object {
    "fullPath": Array [],
    "parentPath": Array [],
    "path": Array [],
    "type": "object",
  },
  Object {
    "fullPath": Array [
      "properties",
      "orderNumber",
    ],
    "parentPath": Array [],
    "path": Array [
      "properties",
      "orderNumber",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "orderDate",
    ],
    "parentPath": Array [],
    "path": Array [
      "properties",
      "orderDate",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "customer",
    ],
    "parentPath": Array [],
    "path": Array [
      "properties",
      "customer",
    ],
    "type": "object",
  },
  Object {
    "fullPath": Array [
      "properties",
      "customer",
      "properties",
      "customerNumber",
    ],
    "parentPath": Array [
      "properties",
      "customer",
    ],
    "path": Array [
      "properties",
      "customerNumber",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "customer",
      "properties",
      "firstName",
    ],
    "parentPath": Array [
      "properties",
      "customer",
    ],
    "path": Array [
      "properties",
      "firstName",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "customer",
      "properties",
      "lastName",
    ],
    "parentPath": Array [
      "properties",
      "customer",
    ],
    "path": Array [
      "properties",
      "lastName",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "orderItems",
    ],
    "parentPath": Array [],
    "path": Array [
      "properties",
      "orderItems",
    ],
    "type": "array",
  },
  Object {
    "fullPath": Array [
      "properties",
      "orderItems",
      "items",
    ],
    "parentPath": Array [
      "properties",
      "orderItems",
    ],
    "path": Array [
      "items",
    ],
    "type": "object",
  },
  Object {
    "fullPath": Array [
      "properties",
      "orderItems",
      "items",
      "properties",
      "productNumber",
    ],
    "parentPath": Array [
      "properties",
      "orderItems",
      "items",
    ],
    "path": Array [
      "properties",
      "productNumber",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "orderItems",
      "items",
      "properties",
      "quantity",
    ],
    "parentPath": Array [
      "properties",
      "orderItems",
      "items",
    ],
    "path": Array [
      "properties",
      "quantity",
    ],
    "type": "integer",
  },
  Object {
    "fullPath": Array [
      "properties",
      "orderItems",
      "items",
      "properties",
      "unitPrice",
    ],
    "parentPath": Array [
      "properties",
      "orderItems",
      "items",
    ],
    "path": Array [
      "properties",
      "unitPrice",
    ],
    "type": "integer",
  },
  Object {
    "fullPath": Array [
      "properties",
      "shippingAddress",
    ],
    "parentPath": Array [],
    "path": Array [
      "properties",
      "shippingAddress",
    ],
    "type": "object",
  },
  Object {
    "fullPath": Array [
      "properties",
      "shippingAddress",
      "properties",
      "street",
    ],
    "parentPath": Array [
      "properties",
      "shippingAddress",
    ],
    "path": Array [
      "properties",
      "street",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "shippingAddress",
      "properties",
      "city",
    ],
    "parentPath": Array [
      "properties",
      "shippingAddress",
    ],
    "path": Array [
      "properties",
      "city",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "shippingAddress",
      "properties",
      "country",
    ],
    "parentPath": Array [
      "properties",
      "shippingAddress",
    ],
    "path": Array [
      "properties",
      "country",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "shippingAddress",
      "properties",
      "zip",
    ],
    "parentPath": Array [
      "properties",
      "shippingAddress",
    ],
    "path": Array [
      "properties",
      "zip",
    ],
    "type": "string",
  },
  Object {
    "fullPath": Array [
      "properties",
      "totalPrice",
    ],
    "parentPath": Array [],
    "path": Array [
      "properties",
      "totalPrice",
    ],
    "type": "number",
  },
]
`;
